src/devices/machine/spifi3.cpp

```diff
@@ -0,0 +1,1727 @@
+// license:BSD-3-Clause
+// copyright-holders:Brice Onken,Olivier Galibert
+
+/*
+ * HP 1TV3-0302 SPIFI3-SE SCSI controller
+ *
+ * References:
+ * - https://github.com/NetBSD/src/blob/trunk/sys/arch/newsmips/apbus/spifireg.h
+ * - https://github.com/NetBSD/src/blob/trunk/sys/arch/newsmips/apbus/spifi.c
+ * - https://github.com/mamedev/mame/blob/master/src/devices/machine/ncr5390.cpp
+ *
+ * TODO:
+ *  - NetBSD compatibility
+ *  - Reselection, target mode, SDTR
+ *  - LUN selection (currently assumes 0)
+ *  - Non-chip-reset conditions
+ *  - Other SPSTAT and ICOND values
+ *  - CMDPAGE details
+ *  - Anything the Sony NEWS driver doesn't use
+ */
+
+#include "emu.h"
+#include "spifi3.h"
+
+#define LOG_GENERAL (1U << 0)
+#define LOG_STATE (1U << 1)
+#define LOG_INTERRUPT (1U << 2)
+#define LOG_DATA (1U << 3)
+#define LOG_REGISTER (1U << 4)
+#define LOG_CMD (1U << 5)
+#define LOG_AUTO (1U << 6)
+
+#define SPIFI3_DEBUG (LOG_GENERAL | LOG_REGISTER | LOG_INTERRUPT | LOG_AUTO)
+#define SPIFI3_TRACE (SPIFI3_DEBUG | LOG_STATE | LOG_CMD)
+#define SPIFI3_MAX (SPIFI3_TRACE | LOG_DATA)
+
+#include "logmacro.h"
+
+DEFINE_DEVICE_TYPE(SPIFI3, spifi3_device, "spifi3", "HP 1TV3-0302 SPIFI3 SCSI-2 Protocol Controller")
+
+spifi3_device::spifi3_device(machine_config const &mconfig, char const *tag, device_t *owner, u32 clock)
+	: nscsi_device(mconfig, SPIFI3, tag, owner, clock),
+	  nscsi_slot_card_interface(mconfig, *this, DEVICE_SELF),
+	  m_irq_handler(*this),
+	  m_drq_handler(*this)
+{
+}
+
+void spifi3_device::device_start()
+{
+	nscsi_device::device_start();
+
+	m_irq_handler.resolve_safe();
+	m_drq_handler.resolve_safe();
+
+	bus_id = 0;
+	tm = timer_alloc(0);
+}
+
+void spifi3_device::map(address_map &map)
```

Comment:
I’m not really fond of the “half the code in the address map” style – it makes the memory layout harder to see.

Reply:
Should be better now.